File: python/state_recover.py
import numpy as np

def centered_mod(v, q):
    v = np.mod(v, q)
    half_q = q // 2
    return np.where(v > half_q, v - q, v)

def state_recover(Vx, scale=1.0, q=None):
    if q is not None:
        Vx = centered_mod(Vx, q)
    return Vx / scale

File: matlab/state_recover.m
function x_rec = state_recover(Vx, scale, q)
    half_q = floor(q/2);
    x_centered = Vx;

    idx = x_centered > half_q;
    x_centered(idx) = x_centered(idx) - q;

    x_rec = x_centered / scale;
end
